src/main/java/edu/harvard/iq/dataverse/importer/metadata/mxrdr/CbfImporter.java

```diff
@@ -0,0 +1,182 @@
+        this.registry = registry;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    @PostConstruct
+    public void init() {
+        registry.register(this);
+    }
+
+    @Override
+    public String getMetadataBlockName() {
+        return "macromolecularcrystallography";
+    }
+
+    @Override
+    public ResourceBundle getBundle(Locale locale) {
+        return ResourceBundle.getBundle("CbfImporterBundle", locale);
+    }
+
+    @Override
+    public ImporterData getImporterData() {
+        return new ImporterData()
+                .addDescription("cbf.modal.description")
+                .addField(ImporterData.ImporterField.of(CbfImporterForm.CBF_FILE, ImporterFieldType.UPLOAD_TEMP_FILE,
+                                                        true, getBundle(Locale.ENGLISH).getString("cbf.label"),
+                                                        ""));
+    }
+
+    @Override
+    public List<ResultField> fetchMetadata(Map<ImporterFieldKey, Object> map) {
+
+        if (map.containsKey(CbfImporterForm.CBF_FILE)) {
+            File cbfFile = (File) map.get(CbfImporterForm.CBF_FILE);
+
+            List<String> cbfLines = readCbfWithoutBinaryData(cbfFile);
+
+            return extractMetadataFields(cbfLines);
+        }
+
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Map<ImporterFieldKey, String> validate(Map<ImporterFieldKey, Object> importerInput) {
+        Map<ImporterFieldKey, String> errors = new HashMap<>();
+
+        File cbfFile = (File) importerInput.get(CbfImporterForm.CBF_FILE);
+
+        if (cbfFile != null && !FilenameUtils.getExtension(cbfFile.getName()).equals("cbf")) {
+            errors.put(CbfImporterForm.CBF_FILE, getBundle(Locale.ENGLISH).getString("cbf.error.wrongFile"));
+            return errors;
+        }
+
+        return errors;
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    /**
+     * Reads headers from cbf file while skipping binary data which is very long in terms of text lines.
+     */
+    private List<String> readCbfWithoutBinaryData(File cbfFile) {
+        List<String> lines = new ArrayList<>();
+
+        try (BufferedReader reader = Files.newBufferedReader(cbfFile.toPath(), Charset.forName("windows-1252"))) {
+
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (line.contains("CIF-BINARY-FORMAT-SECTION")) {
+                    break;
+                }
+                lines.add(line);
+            }
+
+        } catch (IOException e) {
+           throw new IllegalStateException("There was a problem with reading cbf file", e);
+        }
+
+        return lines;
+    }
+
+    private List<ResultField> extractMetadataFields(List<String> cbfLines) {
+        List<MetadataField> metadataFilters = new CbfMetadataFields().getMetadataFilters();
+        List<ResultField> extractedMetadata = new ArrayList<>();
```

Comment:
Have you considered implementing it in not so generic way? Maybe I'm wrong but I think that we are trying to build here some generic solution for non generic task.
I'm thinking something like adding class that will have results of parsing: `CbfFileParseResults` and instead of CbfMetadataFields with some filters adding class `CbfFileParser` with method `CbfFileParseResults` CbfFileParseResults.parse(List<String>) and in this method just converting `CbfFileParseResults` into `List<ResultField>`.
What do you think about that?

Reply:
Hmm, I can I agree i think it would be simpler solution. One thing I think could be obsolete is "CbfFileParseResults" if we are having method in CbfFileParser we might as well return List<ResultField> since it's not like we are going to use "CbfFileParseResults" anywhere. I kinda liked the idea of having field alongside the filter since it is clear how to extract certain field. But maybe it will be easier to read for others.

Reply:
What do you think?

Reply:
Reason behind additional `CbfFileParseResults` is for very easily distinguishing what actually is possible to parse and for easier reading of code. In addition it would be easier to convert `CbfFileParseResults` to `List<ResultField>`. With `List<>` you will have harder time to convert results into what should be returned as final importer result (List<ResultField>). Or if you would already return `List<ResultField>` then you wouldn't have that problem. I guess it would be ok

Reply:
Maybe I'm missing something here so I will draw the flow. I'm reading the file into List<String> -> I'm using ` List<ResultField> cbfFileParser.parse(List<String>).` Where would be the step to parse it into CbfFileParseResults? Or maybe you mean something like `File cbf -> List<CbfFileParseResults> cbfFileParser.parse(File file) -> List<ResultField> cbfFileParser.parse(List<CbfFileParseResults> )`. Do I understand this correctly? If yes then CbfFileParseResults would have only one field, String.

Reply:
I meant:
 - reading file into list
 - CbfFileParseResult cbfFileParser.parse(List<String>) (return type is not a list, maybe 's was missleading)
 - CbfFileParseResult -> List<ResultField>
Version without CbfFileParseResults
 - reading file into list
 - List<ResultField> cbfFileParser.parse(List<String>)

Reply:
Ok so in 1st version what are the fields of CbfFileParseResult? Since in my head we need to parse it into Tree structure similarly how List<ResultField> looks like. If that's the case then it would be almost 1:1 mapping from CbfFileParseResult to List<ResultField>.

Reply:
Yes. It would be very similar to the structure of how `List<ResultField>` would need to look like, but from I understand you could have dataCollection flatten out (there is only one in cbf? at least I don't see that now it have support for two or more).
```
class CbfFileParseResult
   String detectorType
   String detectorDistance
   String angleIncrement
   String oscillationStepSize
...
```

Reply:
Personally I'm more inclined towards non CbfFileParseResults since after first parsing I would have to use a lot of getters and if's to map names and values. I fell like it would be kinda redundant with parsing for the 1st time since in previous method I would also have to check for field names.

Reply:
ok